### PR TITLE
Imperative, Incremental, « stateless » searchable document schema

### PR DESCRIPTION
**Pull request by rtxm**
Alternative to the declarative schema for runtime built schemas.

### CODE REVIEW

File: kansha/services/search/schema.py [hunk 1]
```diff
@@ -167,3 +192,104 @@ def delta(cls, docid, **fields):
...
+
+class Schema(object):
+    '''Imperatively build a Document schema.
+
+    Same reserved keywords as the Declarative Schema.
+
+    Usage:
+
+        TestDocument = schema.Schema('MyDocument')
+        TestDocument.add_field('title', schema.TEXT(stored=True))
```
Comment (nagareproject) [inline kansha/services/search/schema.py]: "There should be one-- and preferably only one --obvious way to do it."

-> ``TestDocument.add_field(schema.TEXT('title', stored=True))``

File: kansha/services/search/schema.py [hunk 2]
```diff
@@ -167,3 +192,104 @@ def delta(cls, docid, **fields):
...
+        )
+
+    TestDocument is then used as if it was a Declarative Document schema.
+    '''
+
+    def __init__(self, name):
+        self.type_name = name
+        self.fields = {}
+
+    def add_field(self, name, field):
```
Comment (nagareproject) [inline kansha/services/search/schema.py]: "There should be one-- and preferably only one --obvious way to do it."

```python
def add_field(self, name, field):
    assert(name != 'doc_type')
    assert(not hasattr(self, name))

    field.schema = self
    self.fields[name] = field
```

File: kansha/services/search/schema.py [hunk 3]
```diff
@@ -167,3 +192,104 @@ def delta(cls, docid, **fields):
...
+
+        assert(name != 'doc_type')
+        assert(not hasattr(self, name))
+        if isinstance(field, FieldType):
+            field.name = name
+            field.parent = self
```
Comment (nagareproject) [inline kansha/services/search/schema.py]: Too generic name `parent` -> `schema`

File: kansha/services/search/schema.py [hunk 4]
```diff
@@ -167,3 +192,104 @@ def delta(cls, docid, **fields):
...
+            schema.TEXT('tags', stored=False) +
+            schema.INT('pages', stored=True) +
+            schema.TEXT('description') +
+            schema.FLOAT('price', stored=True, indexed=False)
+        )
+
+    TestDocument is then used as if it was a Declarative Document schema.
+    '''
+
+    def __init__(self, name):
```
Comment (nagareproject) [inline kansha/services/search/schema.py]: ``def __init__(self, name, *fields)``

File: kansha/services/search/schema.py [hunk 5]
```diff
@@ -167,3 +192,104 @@ def delta(cls, docid, **fields):
...
+        return AltDocument(self, self.fields, docid, **fields)
+
+    def delta(self, docid, **fields):
+        '''
+        Alternate constructor to build delta documents.
+        Fields values not specified to the constructor are set to None.
+        '''
+        for fname in self.fields:
+            if fname not in fields:
+                fields[fname] = None
```
Comment (nagareproject) [inline kansha/services/search/schema.py]:
```python
for fname in set(self.fields) - set(fields):
    fields[fname] = None
```

File: kansha/services/search/schema.py [hunk 6]
```diff
@@ -32,9 +32,9 @@ class MyDoc(Document):
 class FieldType(object):
 
     parent = None
-    name = ''
 
-    def __init__(self, indexed=True, stored=False, default=None):
+    def __init__(self, name='', indexed=True, stored=False, default=None):
```
Comment (nagareproject) [inline kansha/services/search/schema.py]: mandatory `name`

```python
def __init__(self, name, indexed=True, stored=False, default=None):
```

File: kansha/services/search/elasticengine.py [hunk 7]
```diff
@@ -145,7 +145,7 @@ def _index(self, document, update=False):
         doc = 'doc' if update else '_source'
         op = {
             '_index': self.index,
-            '_type': document.__class__.__name__,
+            '_type': document.doc_type.type_name,
```
Comment (nagareproject) [inline kansha/services/search/elasticengine.py]: Dimeters's law. Let the document generate this entry

File: kansha/services/search/elasticengine.py [hunk 8]
```diff
@@ -145,7 +145,7 @@ def _index(self, document, update=False):
...
             '_op_type': 'update' if update else 'create',
             '_id': document._id,
             doc: {k: getattr(document, k)
```
Comment (nagareproject) [inline kansha/services/search/elasticengine.py]: Dimeters's law. Let the document generate this entry

File: kansha/services/search/elasticengine.py [hunk 9]
```diff
@@ -198,7 +198,7 @@ def search(self, query, size=20):
         '''
         dsl = query(self.mapper)
         hits = self.es.search(index=self.index,
-                              doc_type=query.queried_doc.__name__,
+                              doc_type=query.queried_doc.type_name,
```
Comment (nagareproject) [inline kansha/services/search/elasticengine.py]: Dimeters's law

```python
hits = query.search(self.es, self.index, self.mapper, size)
```
```

File: kansha/services/search/schema.py (hunk 6 above)
Comment (rtxm) [inline kansha/services/search/schema.py]: Didn't want to ditch the Declarative Schema where name is provided by attribute assignement.

Comment (nagareproject) [inline kansha/services/search/schema.py]: Personaly, I'd want to ditch it.

But, in the case we keep it, it must be rewrite "on top" of the imperative "Schema" class. By keeping a `schema` attribute or, if possible, perhaps the `Schema` class itself could be both imperative and declarative

File: kansha/services/search/elasticengine.py (hunk 9 above)
Comment (rtxm) [inline kansha/services/search/elasticengine.py]: We don't want to expose specific internal drivers to query.

**Pull request merged**